File: common/src/main/java/com/evandev/remi/feature/creativemodetab/gui/itemtab/ItemTabNavigationBar.java
package com.evandev.remi.feature.creativemodetab.gui.itemtab;

import com.evandev.remi.config.ReliableEmiConfig;
import com.evandev.remi.feature.creativemodetab.gui.CreativeModeTabGui;
import com.evandev.remi.integration.emi.ScreenManager;
import com.google.common.collect.ImmutableList;
import dev.emi.emi.config.SidebarTheme;
import dev.emi.emi.screen.EmiScreenManager;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.AbstractContainerWidget;
import net.minecraft.client.gui.components.TabButton;
import net.minecraft.client.gui.components.events.GuiEventListener;
import net.minecraft.client.gui.layouts.GridLayout;
import net.minecraft.client.gui.narration.NarrationElementOutput;
import net.minecraft.network.chat.Component;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class ItemTabNavigationBar extends AbstractContainerWidget {
    private final ItemTabManager tabManager;
    private final boolean isVertical;
    private final boolean isRightSide;
    public List<ItemTabButton> tabButtons = new ArrayList<>();
    public List<ItemTab> visibleTabs = new ArrayList<>();
    private GridLayout layout = new GridLayout();

    public ItemTabNavigationBar(ItemTabManager tabManager, boolean isVertical, boolean isRightSide) {
        super(0, 0, 0, 0, Component.empty());
        this.tabManager = tabManager;
        this.isVertical = isVertical;
        this.isRightSide = isRightSide;
    }

    public void pos(int x, int y) {
        this.setX(x);
        this.setY(y);
        arrangeElements();
    }

    public void setTabs(List<ItemTab> tabs) {
        this.visibleTabs = new ArrayList<>(tabs);
        if (tabs.isEmpty()) {
            this.tabButtons = ImmutableList.of();
            this.layout = new GridLayout();
            arrangeElements();
            return;
        }
        GridLayout newLayout = new GridLayout();
        newLayout.defaultCellSetting().padding(0);
        EmiScreenManager.SidebarPanel panel = ScreenManager.getTargetCreativeTabPanel();
        SidebarTheme panelTheme = panel != null ? panel.theme : SidebarTheme.TRANSPARENT;
        ImmutableList.Builder<ItemTabButton> buttonBuilder = ImmutableList.builder();

        int leftoverHeight = height % tabs.size();
        int leftoverWidth = width % tabs.size();

        for (int i = 0; i < tabs.size(); i++) {
            ItemTab tab = tabs.get(i);
            ItemTabButton.ButtonStyle buttonStyle = !isVertical ? ItemTabButton.ButtonStyle.TOP
                    : isRightSide ? ItemTabButton.ButtonStyle.RIGHT : ItemTabButton.ButtonStyle.LEFT;
            int w = isVertical ? ReliableEmiConfig.verticalTabsWidth + (panel != null ? panel.theme.horizontalPadding : 0) : ReliableEmiConfig.horizontalTabsWidth;
            int h = isVertical ? ReliableEmiConfig.verticalTabsHeight : ReliableEmiConfig.horizontalTabsHeight + (panel != null ? panel.theme.verticalPadding : 0);

            if (ReliableEmiConfig.tabAlignment == CreativeModeTabGui.TabAlignment.STRETCH) {
                if (isVertical) {
                    h = height / tabs.size();
                    if (leftoverHeight > 0) {
                        h += 1;
                        leftoverHeight--;
                    }
                } else {
                    w = width / tabs.size();
                    if (leftoverWidth > 0) {
                        w += 1;
                        leftoverWidth--;
                    }
                }
            }

            ItemTabButton.TabPosition tabPosition = ItemTabButton.TabPosition.MIDDLE;

            // Soooo many conditions. This could be nicer but I'm tired and it works!
            if (!CreativeModeTabGui.showScrollButtons || panelTheme != SidebarTheme.VANILLA) {
                if (i == 0 && (panelTheme != SidebarTheme.VANILLA || ReliableEmiConfig.tabAlignment == CreativeModeTabGui.TabAlignment.STRETCH || ReliableEmiConfig.tabAlignment == CreativeModeTabGui.TabAlignment.START)) {
                    tabPosition = ItemTabButton.TabPosition.FIRST;
                }
                if (i == tabs.size() - 1 && (panelTheme != SidebarTheme.VANILLA || ReliableEmiConfig.tabAlignment == CreativeModeTabGui.TabAlignment.STRETCH || ReliableEmiConfig.tabAlignment == CreativeModeTabGui.TabAlignment.END)) {
                    tabPosition = ItemTabButton.TabPosition.LAST;
                }
            }

            ItemTabButton button = new ItemTabButton(tabManager, tab, w, h, buttonStyle, tabPosition);
            buttonBuilder.add(button);
            if (isVertical) newLayout.addChild(button, i, 0);
            else newLayout.addChild(button, 0, i);
        }
        this.tabButtons = buttonBuilder.build();
        this.layout = newLayout;
        arrangeElements();
    }

    public void arrangeElements() {
        EmiScreenManager.SidebarPanel panel = ScreenManager.getTargetCreativeTabPanel();

        layout.setX(getX());
        layout.setY(getY());
        layout.arrangeElements();
        if (isVertical) {
            width = ReliableEmiConfig.verticalTabsWidth + (panel != null ? panel.theme.horizontalPadding : 0);
        } else {
            height = ReliableEmiConfig.horizontalTabsHeight + (panel != null ? panel.theme.verticalPadding : 0);
        }
    }

    @Override
    public List<? extends GuiEventListener> children() {
        return this.tabButtons;
    }

    @Override
    public void renderWidget(@NotNull GuiGraphics raw, int mouseX, int mouseY, float partialTick) {
        if (EmiScreenManager.isDisabled()) return;
        tabButtons.forEach(b -> b.render(raw, mouseX, mouseY, partialTick));
    }

    @Override
    public void setFocused(@Nullable GuiEventListener child) {
        super.setFocused(child);
        if (child instanceof TabButton tb) {
            tabManager.setCurrentTab(tb.tab(), false);
        }
    }

    public void setFocusedChild(GuiEventListener child) {
        setFocused(child);
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        for (TabButton child : tabButtons) {
            if (child.mouseClicked(mouseX, mouseY, button)) {
                setFocused(child);
                return true;
            }
        }
        return super.mouseClicked(mouseX, mouseY, button);
    }

    @Override
    public void updateWidgetNarration(@NotNull NarrationElementOutput narrationElementOutput) {
    }
}

File: common/src/main/kotlin/concerrox/emixx/content/creativemodetab/gui/itemtab/ItemTabNavigationBar.kt
package concerrox.emixx.content.creativemodetab.gui.itemtab

import com.google.common.collect.ImmutableList
import com.mojang.blaze3d.systems.RenderSystem
import concerrox.emixx.content.ScreenManager.ENTRY_SIZE
import concerrox.emixx.res
import dev.emi.emi.runtime.EmiDrawContext
import dev.emi.emi.screen.EmiScreenManager
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.AbstractWidget
import net.minecraft.client.gui.components.TabButton
import net.minecraft.client.gui.components.events.GuiEventListener
import net.minecraft.client.gui.layouts.GridLayout
import net.minecraft.client.gui.narration.NarrationElementOutput
import net.minecraft.network.chat.Component

class ItemTabNavigationBar(
    private val tabManager: ItemTabManager,
    private val isVertical: Boolean = false,
    private val isRightSide: Boolean = false
) : AbstractWidget(0, 0, 0, 0, Component.empty()) {

    companion object {
        private val TEXTURE = res("textures/gui/buttons.png")
    }

    private var layout = GridLayout()

    internal var tabButtons: List<TabButton> = emptyList()

    private var focusedChild: GuiEventListener? = null

    var visibleTabs: List<ItemTab> = emptyList()
        private set

    fun pos(x: Int, y: Int): ItemTabNavigationBar {
        this.x = x
        this.y = y
        this.arrangeElements()
        return this
    }

    fun setTabs(tabs: MutableList<ItemTab>) {
        this.visibleTabs = tabs

        val newLayout = GridLayout()
        newLayout.defaultCellSetting().padding(0)

        val buttonBuilder = ImmutableList.builder<TabButton>()

        tabs.forEachIndexed { index, tab ->
            val buttonStyle = if (!isVertical) ItemTabButton.ButtonStyle.TOP
            else if (isRightSide) ItemTabButton.ButtonStyle.RIGHT
            else ItemTabButton.ButtonStyle.LEFT

            val width = if(isVertical) 35 else ENTRY_SIZE
            val height = if(isVertical) 27 else ENTRY_SIZE

            val button = ItemTabButton(tabManager, tab, width, height, buttonStyle, index == 0)
            buttonBuilder.add(button)

            if (isVertical) {
                newLayout.addChild(button, index, 0)
            } else {
                newLayout.addChild(button, 0, index)
            }
        }

        this.tabButtons = buttonBuilder.build()
        this.layout = newLayout

        arrangeElements()
    }

    fun arrangeElements() {
        if (!isVertical) {
            tabButtons.forEach { it.width = ENTRY_SIZE }
        }

        layout.x = if (!isVertical) this.x + 2 else this.x
        layout.y = this.y
        layout.arrangeElements()

        if (isVertical) {
            this.width = 35
            setWidgetHeight(layout.height)
        } else {
            this.width = layout.width + 4
            this.height = layout.height
        }
    }

    private fun setWidgetHeight(newHeight: Int) {
        try {
            val field = AbstractWidget::class.java.getDeclaredField("height")
            field.isAccessible = true
            field.setInt(this, newHeight)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun renderWidget(raw: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        if (EmiScreenManager.isDisabled()) return

        if (!isVertical) {
            RenderSystem.enableBlend()
            EmiDrawContext.wrap(raw).apply {
                drawTexture(TEXTURE, x, y + 2, 32, 0, 1, 16)
                drawTexture(TEXTURE, x + 1, y + 2, 32, 0, 1, 16)
                drawTexture(TEXTURE, x + width - 1, y + 2, 32, 0, 1, 16)
                drawTexture(TEXTURE, x + width - 2, y + 2, 32, 0, 1, 16)
            }
            RenderSystem.disableBlend()
        }

        tabButtons.forEach { it.render(raw, mouseX, mouseY, partialTick) }
    }

    fun setFocusedChild(child: GuiEventListener?) {
        focusedChild?.isFocused = false

        if (child != null) {
            child.isFocused = true
            focusedChild = child
            if (child is TabButton) {
                tabManager.setCurrentTab(child.tab(), false)
            }
        } else {
            focusedChild = null
        }
    }

    override fun setFocused(focused: Boolean) {
        super.setFocused(focused)
        if (!focused) {
            setFocusedChild(null)
        }
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        for (child in tabButtons) {
            if (child.mouseClicked(mouseX, mouseY, button)) {
                setFocusedChild(child)
                return true
            }
        }
        return super.mouseClicked(mouseX, mouseY, button)
    }

    override fun updateWidgetNarration(narrationElementOutput: NarrationElementOutput) {}
}